CHIA-2381 Validate fast forward spends before adding their spend bundle to the mempool

AmineKhaldi opened the pull request:
### Purpose:

Make sure that all fast forward spends of a spend bundle would still have unspent coins.

### Current Behavior:

We can accept a spend bundle with fast forward spends that do not currently have latest unspent coins.

### New Behavior:

We reject spend bundles like that with `DOUBLE_SPEND` error.

arvidn commented on chia/_tests/core/mempool/test_mempool_manager.py:
```diff
@@ -103,6 +103,10 @@ async def zero_calls_get_coin_records(coin_ids: Collection[bytes32]) -> list[Coi
 
 
+async def zero_calls_get_unspent_lineage_info_for_puzzle_hash(_puzzle_hash: bytes32) -> Optional[UnspentLineageInfo]:
+    assert False  # pragma no cover
```
if the test expects this to never be called, I think you should put an `assert False` here, and maybe with a comment explaining why we don't expect it to be called

AmineKhaldi replied:
This function already does `assert False` and it's used in `instantiate_mempool_manager` to conveniently omit it when a test calls `instantiate_mempool_manager` with no need to call that function.

arvidn replied:
oh, heh. I must have been blinded by "pragma no cover"

arvidn commented on chia/types/eligible_coin_spends.py:
```diff
@@ -24,6 +24,7 @@ class EligibilityAndAdditions:
     is_eligible_for_dedup: bool
     spend_additions: list[Coin]
     is_eligible_for_ff: bool
+    ff_puzzle_hash: Optional[bytes32] = None
```
would you mind adding a comment here? I imagine if `is_eligible_for_ff` is `True`, this is not allowed to be `None` and vice versa.

So, technically, the `is_eligible_for_ff` could also be changed from `bool` -> `Optional[bytes32]`.

AmineKhaldi replied:
That's reasonable.

arvidn commented on chia/full_node/mempool_manager.py:
```diff
@@ -435,12 +442,18 @@ async def validate_spend_bundle(
             )
-            mark_as_fast_forward = eligibility_info.is_eligible_for_ff and supports_fast_forward(coin_spend)
+            mark_as_fast_forward = (
+                eligibility_info.is_eligible_for_ff
+                and eligibility_info.ff_puzzle_hash is not None
```
this suggests that `ff_puzzle_hash` may be disconnected from `is_eligible_for_ff`. I wouldn't have thought so

arvidn commented on chia/full_node/mempool_manager.py:
```diff
@@ -452,6 +465,11 @@ async def validate_spend_bundle(
             return Err.INVALID_SPEND_BUNDLE, None, []
 
+        # Make sure all fast forward spends still have unspent coins
+        for puzzle_hash in fast_forward_puzzle_hashes:
+            if await get_unspent_lineage_info_for_puzzle_hash(puzzle_hash) is None:
+                return Err.DOUBLE_SPEND, None, []
```
it's not obvious to me why you defer this check. instead of storing the puzzle hashes to check in `fast_forward_puzzle_hashes` and then checking them, why not just check them directly in the loop above?

arvidn commented on chia/full_node/mempool_manager.py:
```diff
@@ -424,9 +429,11 @@ async def validate_spend_bundle(
                 is_eligible_for_dedup=is_eligible_for_dedup,
                 spend_additions=spend_additions,
                 is_eligible_for_ff=is_eligible_for_ff,
+                ff_puzzle_hash=bytes32(spend.puzzle_hash),
```
do you set the `ff_puzzle_hash` unconditionally here? I would have thought you only set it for ff-spends.

arvidn reviewed (commented):
my main comment is around the relationship of `is_eligible_for_ff` and `ff_puzzle_hash`. It seems like it could be stricter, or maybe merged into a single field

arvidn commented on chia/types/eligible_coin_spends.py:
```diff
@@ -23,7 +23,8 @@
     is_eligible_for_dedup: bool
     spend_additions: list[Coin]
-    is_eligible_for_ff: bool
+    # This is None if the spend is not eligible for fast forward
```
"otherwise it is..."?
It's the puzzle hash of the coin being spent, right?
I think it would also be good to say that this being set means this spend is eligible for fast forward and that we use this to check if the singleton has an unspent coin or not

AmineKhaldi replied:
Yep, I elaborated accordingly in the comment.

arvidn commented on chia/_tests/core/mempool/test_mempool_manager.py:
```diff
@@ -427,18 +436,18 @@ def make_bundle_spends_map_and_fee(
             is_eligible_for_dedup=bool(spend.flags & ELIGIBLE_FOR_DEDUP),
             spend_additions=spend_additions,
-            is_eligible_for_ff=bool(spend.flags & ELIGIBLE_FOR_FF),
+            ff_puzzle_hash=bytes32(spend.puzzle_hash) if bool(spend.flags & ELIGIBLE_FOR_FF) else None,
```
what type does `spend.puzzle_hash` have? is it not `bytes32`?

AmineKhaldi replied:
It's `bytes`: https://github.com/Chia-Network/chia_rs/blob/38d4f845705b1fd9cc68077a9d50a29afd79e6d5/wheel/python/chia_rs/chia_rs.pyi#L269

arvidn replied:
I see. that's a problem we should address separately

arvidn commented on chia/_tests/core/mempool/test_mempool_manager.py:
```diff
@@ -139,8 +143,13 @@ async def instantiate_mempool_manager(
 ) -> MempoolManager:
-    mempool_manager = MempoolManager(get_coin_records, constants, max_tx_clvm_cost=max_tx_clvm_cost)
+    mempool_manager = MempoolManager(
+        get_coin_records, get_unspent_lineage_info_for_puzzle_hash, constants, max_tx_clvm_cost=max_tx_clvm_cost
```
why not just pass `zero_calls_get_unspent_lineage_info_for_puzzle_hash` here directly?

AmineKhaldi replied:
Hmm I didn't want to alter all `instantiate_mempool_manager` calls and increase the patch size. My rationale being if any caller needs an explicit function here, it can pass one.

arvidn replied:
I don't see you alter *any* call to `instantiate_mempool_manager()`. are you?
as far as I can tell, all calls to `instantiate_mempool_manager()` will currently use the `zero_calls_get_unspent_lineage_info_for_puzzle_hash`. Is that not the case?

If you would remove the parameter you added, and pass in `zero_calls_get_unspent_lineage_info_for_puzzle_hash` unconditionally, that would have the same effect, wouldn't it?

arvidn commented on chia/_tests/core/mempool/test_mempool_manager.py:
```diff
@@ -139,8 +143,13 @@ async def instantiate_mempool_manager(
     block_timestamp: uint64 = TEST_TIMESTAMP,
     constants: ConsensusConstants = DEFAULT_CONSTANTS,
     max_tx_clvm_cost: Optional[uint64] = None,
+    get_unspent_lineage_info_for_puzzle_hash: Callable[
+        [bytes32], Awaitable[Optional[UnspentLineageInfo]]
+    ] = zero_calls_get_unspent_lineage_info_for_puzzle_hash,
```
this doesn't look necessary

AmineKhaldi replied:
Please see my rationale above and let me know what you think.

emlowe commented on benchmarks/mempool-long-lived.py:
```diff
@@ -90,9 +91,15 @@ async def get_coin_record(coin_ids: Collection[bytes32]) -> list[CoinRecord]:
         return ret
 
+    # We currently don't need to keep track of these for our purpose
+    async def get_unspent_lineage_info_for_puzzle_hash(_: bytes32) -> Optional[UnspentLineageInfo]:
```
just curious why you didn't also name this `zero_calls_...` like `test_mempool_manager`

AmineKhaldi replied:
Hmm, I guess in consistency with `get_coin_record` right above it, considering this is not production code/tests, I just went with that. Notice I did the same thing with the one in `benchmarks/mempool.py` too.

AmineKhaldi committed:
Validate fast forward spends before adding their spend bundle to the mempool.

Make sure that all fast forward spends of a spend bundle would still have unspent coins.
